File: apps/web/src/lib/api/skills.ts
/** Skills API — wraps /api/v1/skills/*. */
const API_BASE = process.env.NEXT_PUBLIC_API_BASE_URL || 'http://localhost:8000';

export interface SkillRow {
  id: string;
  name: string;
  version: string;
  source_type: 'builtin' | 'local' | 'url' | 'git' | 'mcp_only';
  source_url: string | null;
  status: 'active' | 'disabled' | 'failed' | 'uninstalled' | 'installing';
  scope: 'global' | 'project';
  project_id: string | null;
  manifest_json: Record<string, unknown>;
}

export async function listSkills(params: { scope?: string; project_id?: string } = {}): Promise<SkillRow[]> {
  const qs = new URLSearchParams();
  if (params.scope) qs.set('scope', params.scope);
  if (params.project_id) qs.set('project_id', params.project_id);
  const r = await fetch(`${API_BASE}/api/v1/skills${qs.toString() ? `?${qs}` : ''}`);
  if (!r.ok) throw new Error(`listSkills: ${r.status}`);
  return r.json();
}

export async function installSkill(payload: {
  source_type: 'local' | 'url' | 'git';
  source: string;
  scope: 'global' | 'project';
  project_id?: string | null;
}): Promise<SkillRow> {
  const r = await fetch(`${API_BASE}/api/v1/skills/install`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });
  if (!r.ok) throw new Error(`installSkill: ${r.status} ${await r.text()}`);
  return r.json();
}

export async function enableSkill(skillId: string): Promise<SkillRow> {
  const r = await fetch(`${API_BASE}/api/v1/skills/${skillId}/enable`, { method: 'POST' });
  if (!r.ok) throw new Error(`enableSkill: ${r.status}`);
  return r.json();
}

export async function disableSkill(skillId: string): Promise<SkillRow> {
  const r = await fetch(`${API_BASE}/api/v1/skills/${skillId}/disable`, { method: 'POST' });
  if (!r.ok) throw new Error(`disableSkill: ${r.status}`);
  return r.json();
}

export async function uninstallSkill(skillId: string): Promise<void> {
  const r = await fetch(`${API_BASE}/api/v1/skills/${skillId}`, { method: 'DELETE' });
  if (!r.ok) throw new Error(`uninstallSkill: ${r.status}`);
}

File: apps/api/app/api/routes/skills.py
"""GET/POST/DELETE /v1/skills/* — skill installation management."""
from typing import Optional, Literal

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.db.database import get_db
from app.models.skill_installation import SkillInstallation
from app.services.agent.skills import installer, lifecycle

router = APIRouter(prefix="/v1/skills", tags=["Skills"])


class SkillOut(BaseModel):
    id: str
    name: str
    version: str
    source_type: str
    source_url: Optional[str]
    status: str
    scope: str
    project_id: Optional[str]
    manifest_json: dict

    class Config:
        from_attributes = True


@router.get("", response_model=list[SkillOut])
def list_skills(
    scope: Optional[str] = None,
    project_id: Optional[str] = None,
    db: Session = Depends(get_db),
):
    q = db.query(SkillInstallation).filter(SkillInstallation.status != "uninstalled")
    if scope:
        q = q.filter(SkillInstallation.scope == scope)
    if project_id:
        q = q.filter(SkillInstallation.project_id == project_id)
    return q.order_by(SkillInstallation.installed_at.desc()).all()


class InstallRequest(BaseModel):
    source_type: Literal["local", "url", "git"]
    source: str
    scope: Literal["global", "project"] = "project"
    project_id: Optional[str] = None


@router.post("/install", response_model=SkillOut)
def install(req: InstallRequest, db: Session = Depends(get_db)):
    try:
        if req.source_type == "local":
            inst = installer.install_local(db, req.source, scope=req.scope, project_id=req.project_id)
        elif req.source_type == "url":
            inst = installer.install_url(db, req.source, scope=req.scope, project_id=req.project_id)
        elif req.source_type == "git":
            inst = installer.install_git(db, req.source, scope=req.scope, project_id=req.project_id)
    except Exception as e:
        raise HTTPException(400, str(e))
    return inst


@router.post("/{skill_id}/enable", response_model=SkillOut)
def enable_skill(skill_id: str, db: Session = Depends(get_db)):
    try:
        return lifecycle.enable(db, skill_id)
    except Exception as e:
        raise HTTPException(400, str(e))


@router.post("/{skill_id}/disable", response_model=SkillOut)
def disable_skill(skill_id: str, db: Session = Depends(get_db)):
    return lifecycle.disable(db, skill_id)


@router.delete("/{skill_id}")
def uninstall_skill(skill_id: str, db: Session = Depends(get_db)):
    lifecycle.uninstall(db, skill_id)
    return {"ok": True}
